mountpoint-s3-client/src/error_metadata.rs

```diff
@@ -0,0 +1,29 @@
+#[derive(Default, Debug, Clone, PartialEq)]
+pub struct ErrorMetadata {
+    pub http_code: Option<i32>,
+    pub s3_error_code: Option<String>,
+    pub s3_error_message: Option<String>,
+    pub error_code: Option<String>,
```

Comment:
Valid observation, those fields are never set in the client.

One approach we could take is:
```
// mountpoint-s3-client
pub struct ClientErrorMetadata {
    pub http_code: Option<i32>,
    pub s3_error_code: Option<String>,
    pub s3_error_message: Option<String>,
}

// mountpoint-s3
pub struct ErrorMetadata {
    pub client_error_meta: ClientErrorMetadata,
    pub error_code: Option<String>,
    pub s3_bucket_name: Option<String>,
    pub s3_object_key: Option<String>,
}
``` 

I think it won't even require a second trait, since only client errors are generic parameters, `InodeError` / `fs::Error` are always used when the type is known

Reply:
just as an alternative, this is how rust SKD achieves extendibility of the meta struct (all of their error's are "client" errors though):

```rust
pub struct ErrorMetadata {
    code: Option<String>,
    message: Option<String>,
    extras: Option<HashMap<&'static str, String>>,
}

const EXTENDED_REQUEST_ID: &str = "s3_extended_request_id";

impl RequestIdExt for ErrorMetadata {
    fn extended_request_id(&self) -> Option<&str> {
        self.extra(EXTENDED_REQUEST_ID)
    }
}
```

We could do the same and leave `ErrorMetadata::extras` empty in client, than fill them just before converting the error to `InodeError` and implement a trait/accessor method for each field:

```rust
impl InodeError {
    fn error_code(&self) -> Option<&str> {}
    fn s3_bucket_name(&self) -> Option<&str> {}
    fn s3_object_key(&self) -> Option<&str> {}
}
```

This looks to be boilerplate though as accessors should be implemented for each fs error (at least for `PrefetchReadError` / `UploadWriteError`).